### PR TITLE
Migrate the Stash to Protobuf

### LINKED ISSUE #17466: [Epic] adapter: improve migration framework
### Spec
The adapter migration framework needs some more thoughtful attention. We want to improve our confidence in the framework that it
a)  will work as intended
b) is able to support all the types of migrations we'd like to make, and
c) will prevent any catalog-breaking changes from being released.

Specifically, we should
1) Start by identifying the set of problems we wish to prevent and objectives of our migration framework, covering both stash and catalog migrations. Right now it's not clear and there's no explicit mandate. This is the most critical and difficult piece of work.
2) Once we have a clear understanding of objectives of our migration framework, compare that to the existing implementation(s) (which span 2+ places in the code).
3) Finally, either
    a) our assessment in step 2 above concludes that the existing implementation works, and we make minor improvements and add documentation
    b) we decide to redesign the migration framework, and kick off a formal design process.

Some ideas/related tickets:
- See contents below - change stash migrations to not use indexes into a vector to figure out where to start
- Integrate stash-debug into CI and/or release process workflows  (some similar ideas are being proposed for compute here https://github.com/MaterializeInc/materialize/issues/17883)
- https://github.com/MaterializeInc/materialize/issues/17824
- https://github.com/MaterializeInc/materialize/issues/14125
- https://github.com/MaterializeInc/materialize/issues/14126
  - This issue has a good overview of our current migration framework, but is not a high priority item itself.
- https://github.com/MaterializeInc/materialize/issues/17236
- https://github.com/MaterializeInc/materialize/issues/18485 (@mjibson  was working on this last I checked)
- https://github.com/MaterializeInc/materialize/issues/11435 (old tracker for initial implementation)

----
### Original ticket content 
We perform migrations in multiple places on envd start (in order):

- Stash migrations [link](https://github.com/MaterializeInc/materialize/blob/e96e7e2030de8eecd4a222d9c48514ddea5afc15/src/adapter/src/catalog/storage.rs#L567), called by envd's `serve` function
- Catalog migrations [link](https://github.com/MaterializeInc/materialize/blob/e96e7e2030de8eecd4a222d9c48514ddea5afc15/src/adapter/src/catalog.rs#L2584), called by Catalog::open, which is called by coord's `serve` function

Stash migrations are indexed, and we know which one we've performed, so we start at the one after and do those. Catalog migrations always run all catalog items. A problem with the existing migrations is that often (always?) need to be around for 1 release, and then can be deleted at some release after that. (A migration might exist in the future where a migration *must* be deleted before the next release.) We can usually cleanly delete the catalog migrations, but the stash migrations must stay around in their indexed place. Keeping around old migrations that won't ever do anything again can be bad because they force some unused data structures or fields to exist that should be removed.

Proposal from @benesch:

> I think just have each migration declare its migration ID rather than inferring it from its position in the array, and leave a comment saying what the next migration ID should be.
> In general we want to make these migrations idempotent if applied twice because if something goes wonky with the release (e.g. right now, v0.42 is at risk, and there’s a possibility we go straight from v0.41 -> v0.43), we might want to do something nonstandard with the version progression.
> So I think each migration carrying its migration ID, plus a comment saying the release at which a migration should/can be removed, is a good start? And then we can do a sweep and remove old migrations periodically, but probably don’t need to be too strict about removing them at exactly the next release?
> If we do write a migration that 100% needs to be removed after exactly one release: yeah enforcing that in code sounds good. But ideally that’s not the common case.

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Motivation

Fixes #17466

This is the final PR for improving Stash migrations. The general approach we're taking is to migrate the Stash from having serialized JSON to serialized protobufs. We define the objects we store in a `.proto` file, every time we want to change this file we must first snapshot it (i.e. copy and paste) and write a migration from the previous version to the new version.

All fields in proto3 are optional by default, without any way to mark them as required. To make working with Stash collections easier, we re-use the existing types in the Catalog, implementing `mz_proto::RustType` for them to do the conversion between proto and Rust. I also updated `TableTransaction`, which is widely used in the catalog, to best encapsulate this conversion. While this introduces a bit of overhead, it decouples our application layer from our storage layer, which gives us some additional flexibility with how data is stored in the Stash.

### Tips for reviewer

It's easiest to review this PR one commit at a time. I did my best to break it down into logical sections, the first being the largest.

1. Implement `mz_proto::RustType` for a bunch of types. 
    1. This is straight converting from one type to another. Previously I had added a `StashType` trait but realized the already existing `RustType` had exactly the same intent, and was better in many ways. So this commit involves deleting `StashType` and implementing `RustType` for those types, as well as a number of new types.
2. Changes the Stash itself and the inner postgres workings to use bytes and protobuf.
3. Migrates the Catalog to use protobuf types.
4. Migrates the storage-controller to use protobuf types.
5. Ports the existing Catalog/Stash migrations to the new framework
6. Updates the `stash-debug` crate to use protobuf types.
    1. Note: making edits with stash-debug are now quite difficult since we can’t take a “generic protobuf” as we could take a “generic json blob”. I was thinking to improve this we could use [`prost-reflect`](https://crates.io/crates/prost-reflect) so we could take an arbitrary string as the name of the collection, and then dynamically prompt the user for the fields in the message. Making this change is out of scope for this PR though.
7. Updates the Stash tests, and makes a few correctness changes to the initialization flow.

### Testing

#19208 introduced the actual migration flow. In that PR there are a number of tests to exercise scale and different errors paths. For this PR I'm more so concerned about how the protos work during normal usage of the Stash. As such, I was planning on running the [nightlies](https://buildkite.com/materialize/nightlies) and [SQL Logic Tests](https://buildkite.com/materialize/sql-logic-tests) on this branch.

### Checklist

- [ ]  This PR has adequate test coverage / QA involvement has been duly considered.
- [ ]  This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc [template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md), or is sufficiently small to not require a design.

- [ ]  This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ]  If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).

- [x]  This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
    - N/a

### CODE REVIEW

File: src/persist-client/Cargo.toml
Comment (danhhz) [inline src/persist-client/Cargo.toml:46]: i would much prefer to not introduce this dependency! let's talk about how we might avoid it, please :)

i think the answer might be that ShardId already has `RustType` impl to serialize it to/from a String, which is the intended way to durably write down ShardIds. the`[u8; 16]` are an internal detail

**Review by danhhz — CHANGES_REQUESTED** [review]

File: src/persist-client/Cargo.toml
Comment (ParkMyCar) [reply src/persist-client/Cargo.toml]: Totally makes sense! I updated the PR and removed the dependency from `persist-client`, using `String` as the protobuf type

Comment (danhhz) [reply src/persist-client/Cargo.toml]: 🙌 thanks!

File: src/stash/protos/objects.proto
Comment (aalexandrov) [inline src/stash/protos/objects.proto:1]: What's the difference between the messages defined in `objects.proto` and `objects_vXX.proto`?

**Review by aalexandrov — APPROVED** [review]
`RustType` implementations look good.

File: src/stash/protos/objects.proto
Comment (ParkMyCar) [reply src/stash/protos/objects.proto:1]: `objects.proto` are the current types we'll deserialize from the Stash, `objects_vXX.proto` are snapshots of how they types existed at the previous versions of the Stash. By snapshotting the types it allows us to do arbitrarily complex migrations, e.g. we could change a `uint64` to an `enum` or `message` if need be.

File: src/adapter/src/catalog/error.rs
Comment (def-) [inline src/adapter/src/catalog/error.rs:139]: This seems unused, might be interesting to test the error case?

File: src/adapter/src/catalog/storage/objects.rs
Comment (def-) [inline src/adapter/src/catalog/storage/objects.rs:365]: Might want to test this

File: src/audit-log/src/lib.rs
Comment (def-) [inline src/audit-log/src/lib.rs:652]: Same argument

File: src/audit-log/src/lib.rs
Comment (def-) [inline src/audit-log/src/lib.rs:707]: Another V1

File: src/stash/src/upgrade.rs
Comment (def-) [inline src/stash/src/upgrade.rs:87]: The Delete block appears not to be tested

File: src/stash/src/upgrade.rs
Comment (def-) [inline src/stash/src/upgrade.rs:162]: Another interesting case that could probably be tested by not having a previous version and thus inserting a new one instead of updating.

**Review by def- — COMMENTED** [review]
This change looks interesting enough to involve QA in. The nightly looks ok on a quick look, apart from the test breakages we also have in main right now: https://buildkite.com/materialize/nightlies/builds/2436

I marked interesting parts of the code which appear untested based on the coverage run: https://buildkite.com/materialize/coverage/builds/115 I hope those are useful for you and reviewers.

**Review by jkosh44 — COMMENTED** [review]
I think we should add a README somewhere that describes the full process of how to add new migrations (unless we already did that in a previous PR).

Also, something I was thinking about. We will probably want to be able to correlate stash versions to the actual Materialize version. Is that possible?

File: src/stash/src/postgres.rs
Comment (mjibson) [inline src/stash/src/postgres.rs:903]: Although I think this generally works, I'm a bit scared that we have lots of code that will append empty batches during peek (that is, many reads can also cause writes), and could potentially mess this up? Can we do something else like assert that `iter[_raw]` over this collection is empty? That seems safer than relying on a specific timestamp.

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [inline src/adapter/src/catalog/storage.rs:205]: Does this readonly check need to be here? I think we should complain loudly if this is the case instead of silently moving on, which is almost certainly wrong (if upgrade needed to happen, then all the dependent types assume that, so not running it will produce a data mismatch between on disk and the binary).

**Review by mjibson — APPROVED** [review]
Looking really good!

Although migrations are getting fearless and will produce lots of compiler errors, we should probably iterate a bit more in followup PRs to improve the ease of doing that. A readme as suggested certainly required sometime. A script that could do a bit of the annoying work would be great too.

File: src/adapter/src/catalog/storage/stash.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage/stash.rs:67]: This wasn't used anywhere yet. To make sure the new initialization code did the same thing as the old code, what I ended up doing was writing a snapshot test using [`insta`](https://insta.rs/), snapshotting the debug repr of a `catalog::Transaction`, and then running this branch against that snapshot. 

I didn't add that test to this PR because it was >1000 lines since it includes the snapshot, but I plan on putting it up after this merges if we think there's value.

File: src/stash/protos/objects.proto
Comment (ParkMyCar) [reply src/stash/protos/objects.proto:173]: According to Dan `String` is how we durably persist `ShardId`s. We also define `String` as the protobuf type [for ShardId](https://github.com/MaterializeInc/materialize/blob/b2e4a84ab38416d3ac44a2f2b8410d86a8d1d81c/src/persist-client/src/internal/encoding.rs#L120) and it's what persist uses for this [proto definitions](https://github.com/MaterializeInc/materialize/blob/b2e4a84ab38416d3ac44a2f2b8410d86a8d1d81c/src/persist-client/src/internal/state.proto#L85), so I opted to go for the same thing here.

File: src/stash/src/transaction.rs
Comment (ParkMyCar) [reply src/stash/src/transaction.rs:247]: 😁

File: src/stash/protos/objects.proto
Comment (danhhz) [reply src/stash/protos/objects.proto:173]: yup, confirming that ShardIds are meant to only ever be recorded or sent across the network as the string representation. we intentionally hide the bytes as an internal implementation detail

File: src/adapter/src/catalog/storage.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage.rs:205]: I kept the read_only check because it exists today, but you're right, if we need to upgrade and we don't then chances are we'll hit some panic for a data mismatch. 

I got rid of this check and in the upgrade path we'll check if it's read only and we need to upgrade, if so we'll log an error to Sentry.

File: src/adapter/src/catalog/storage.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage.rs:232]: Did the same as above for these migrations^

File: src/adapter/src/catalog/error.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog/error.rs:139]: Let me come back to this, I'm pretty sure it's used in the implementation of `TableTransaction` which is part of `mz_catalog::Transaction`, so I would expect it to be tested 🤔

File: src/adapter/src/catalog/storage/objects.rs
Comment (ParkMyCar) [reply src/adapter/src/catalog/storage/objects.rs:365]: Currently there isn't a way to create a "system database". I'm thinking about adding some more randomized testing with proptest, do you mind if I come back to this case?

File: src/audit-log/src/lib.rs
Comment (ParkMyCar) [reply src/audit-log/src/lib.rs:499]: I derived `proptest::Arbitrary` for all of the audit-log and storage usage types, and made there their proto implementations round tripped through serialized. Is that enough, or should we add more coverage?

File: src/audit-log/src/lib.rs
Comment (ParkMyCar) [reply src/audit-log/src/lib.rs:653]: ^ added proptest for this type

File: src/audit-log/src/lib.rs
Comment (ParkMyCar) [reply src/audit-log/src/lib.rs:708]: ^ added proptest for this type

File: src/stash/src/upgrade.rs
Comment (ParkMyCar) [reply src/stash/src/upgrade.rs:117]: No migrations need to delete keys yet, since this PR is so large do you mind if I follow up with this?

**Comment by ParkMyCar** [comment]
> This change looks interesting enough to involve QA in. The nightly looks ok on a quick look, apart from the test breakages we also have in main right now: https://buildkite.com/materialize/nightlies/builds/2436
> 
> I marked interesting parts of the code which appear untested based on the coverage run: https://buildkite.com/materialize/coverage/builds/115 I hope those are useful for you and reviewers.

Thanks for the review and the running the testing coverage @def-!

**Comment by ParkMyCar** [comment]
> I think we should add a README somewhere that describes the full process of how to add new migrations (unless we already did that in a previous PR).

I added a module comment in `stash/src/upgrade.rs` that hopefully adds enough detail as to how Stash migrations work. I'm hoping to put up a PR for #19207 that can act as a guide for folks.


> Also, something I was thinking about. We will probably want to be able to correlate stash versions to the actual Materialize version. Is that possible?

This would probably be possible, it depends on what you mean by correlate, maybe git tags? What reason are you thinking of for correlating versions, maybe for debugging?

File: src/stash/src/upgrade.rs
Comment (ParkMyCar) [reply src/stash/src/upgrade.rs:97]: Yeah I'm pretty sure this is just a single retry. I'm not sure what would cause this to fail, honestly I copied it from other Stash methods, [example](https://github.com/MaterializeInc/materialize/blob/main/src/stash/src/lib.rs#L615-L622).

File: src/stash/src/upgrade/v15_to_v16.rs
Comment (ParkMyCar) [reply src/stash/src/upgrade/v15_to_v16.rs:34]: The hard-coding is actually by design! Say in the future we change the name of the system cluster, or how the `USAGE` flag is represented for `AclMode`. If in the migration path we depended on our common types, then a change like this could break the migration. By hard-coding these values we're effectively freezing the state of the Stash in time, which protects our migrations from our future-selves

**Comment by ParkMyCar** [comment]
> Looking really good!

Thanks!

> Although migrations are getting fearless and will produce lots of compiler errors, we should probably iterate a bit more in followup PRs to improve the ease of doing that. A readme as suggested certainly required sometime. A script that could do a bit of the annoying work would be great too.

I totally agree! I wanted to add more testing or automation (I have a couple WIP branches) but this PR is already so large I figured I would hold off until this one is merged. These other changes should also be much smaller and parallelize-able.

File: src/stash/src/postgres.rs
Comment (ParkMyCar) [reply src/stash/src/postgres.rs]: Makes sense. I updated the impl to first read what collections we have (at a given point the Stash's `collections` member var might not be initialized) and only if the config collection exists do we try to read it. I think this should avoid any writes from occurring? If I missing something though let me know!

File: src/stash/src/postgres.rs
Comment (ParkMyCar) [reply src/stash/src/postgres.rs:956]: Yeah v16 of the Stash didn't change the serialized typed, but we needed to bump the Stash version so we could port [this new migration](https://github.com/ParkMyCar/materialize/blob/c6a75edcf94c5323849386d4434ef3585477a4b4/src/adapter/src/catalog/storage.rs#L534-L558)

File: src/stash/src/postgres.rs
Comment (mjibson) [inline src/stash/src/postgres.rs:900]:
```suggestion
    /// Returns whether this Stash is initialized. We consider a Stash to be initialized if
```

**Review by mjibson — APPROVED** [review]
is_initialized change lgtm

File: src/adapter/src/catalog/storage.rs
Comment (mjibson) [reply src/adapter/src/catalog/storage.rs]: Erp I think I remember why this is needed. Testdrive does a catalog validation on each DDL, and *that* needs to not even try to upgrade things because they'll fail. If it's already there and starts causing test failures, put it back.

File: src/adapter/src/catalog/storage/objects.rs
Comment (def-) [reply src/adapter/src/catalog/storage/objects.rs:365]: No, that's fine. I just thought it might be interesting for you to know which parts are not covered by existing tests.

File: src/adapter/src/catalog/error.rs
Comment (def-) [reply src/adapter/src/catalog/error.rs:139]: Easy way to verify is adding a panic in here and running the test you expect to cover it.

File: src/stash/src/upgrade.rs
Comment (jkosh44) [reply src/stash/src/upgrade.rs:97]: In that case it's probably right? @mjibson would probably know since I'm pretty sure he wrote the linked code.

File: src/stash/src/upgrade/v15_to_v16.rs
Comment (jkosh44) [reply src/stash/src/upgrade/v15_to_v16.rs:34]: Is it possible that when we generate the `objects_vX` structs, that we can generate them with all the same methods/fields that the struct had at the time of the migration? That way we'd be able to call something like `objects_v15::AclMode::USAGE`, and we'd know it give us the result of what `USAGE` would have returned in v15?

I'm just worried that it will be way to easy to accidentally type  something like `bitflags: 1 << 7` and cause havoc.

**Comment by jkosh44** [comment]
> @jkosh44
> 
> > I think we should add a README somewhere that describes the full process of how to add new migrations (unless we already did that in a previous PR).
> 
> I added a module comment in `stash/src/upgrade.rs` that hopefully adds enough detail as to how Stash migrations work. I'm hoping to put up a PR for #19207 that can act as a guide for folks.

Are you sure you linked the right issue here? Or did you just mean that when you fix that issue you are also going to add in a guide?

> > Also, something I was thinking about. We will probably want to be able to correlate stash versions to the actual Materialize version. Is that possible?
> 
> This would probably be possible, it depends on what you mean by correlate, maybe git tags? What reason are you thinking of for correlating versions, maybe for debugging?

For each environment, these stash migrations only get run once. In staging and prod, they will always be run during a Materialize version upgrade. So each stash migration from `objects_vX` to `objects_vY` will correspond to exactly one Materialize version upgrade from `vA` to `vB`. In case something goes wrong and we need to do some debugging or just for audit purposes, it will likely be useful to know what Materialize version upgrade corresponds to each stash migration.

File: src/stash/src/upgrade.rs
Comment (mjibson) [reply src/stash/src/upgrade.rs:97]: Yes, this does a single retry. It is a pattern used in all peek fns, which need a sealed timestamp. At init time when since == upper, there's no sealed timestamp, so the peek fns forcibly advance upper by 1.

**Comment by ParkMyCar** [comment]
Rebased off `main` and pulled in the changes from both https://github.com/MaterializeInc/materialize/commit/c5f5c97509500d258aa5eb1a68d3cbb7d66a513d and https://github.com/MaterializeInc/materialize/commit/988667000e6e62dd5c861a6153355f108b845335

**Comment by ParkMyCar** [comment]
I used an unmerged [snapshot](https://github.com/mitsuhiko/insta) test (https://github.com/MaterializeInc/materialize/compare/main...ParkMyCar:materialize:stash/snapshot-test) to make sure the initial state of the Stash on main is equal to what it is with this migration. So we should we should have picked up any changes that were introduced since I last pushed this branch

**Pull request merged**
